### PR TITLE
[CSClosure] Add support for projected/wrapper values

### LINKED ISSUE #59295: Property Wrapper cannot be compiled correctly with `$` in function body (Xcode 14 and main?)
**Describe the bug**
The code with Property Wrapper cannot be compiled correctly in Swift 5.7 dev (Xcode 14 dev)

```swift
import Foundation

@propertyWrapper
struct WrapperValue<Value> {
  var value: Value
  init(wrappedValue: Value) {
    self.value = wrappedValue
  }

  var projectedValue: Self {
    return self
  }
  
  var wrappedValue: Value {
    get {
      self.value  
    }
    set {
      self.value = newValue
    }
  }

  func printValue() {
    print(value)
  }
}


class Test {
  static func test() {
    (0..<100).map { _ in
      @WrapperValue var value: Bool = false

      if value {
        $value.printValue() // .. compile error on 5.7
      }
    }
  }
}

```

This code had been compiled until Swift 5.6.1.

I already send it to apple (FB10059129)

**To Reproduce**
Compile above code.

**Expected behavior**
Compile success.

**Screenshots**

```
<stdin>:35:9: error: cannot reference invalid declaration '$value'
        $value.printValue() // .. compile error on 5.7
        ^
<stdin>:32:25: note: '$value' declared here
      @WrapperValue var value: Bool = false
                        ^
```

**Environment (please complete the following information):**
 - Xcode 14 dev (14A5228q)
 - nightly (ee312bc1e20eb01)

### PR DESCRIPTION
**Pull request by xedin**
To support referencing projected and/or wrapped var
in a closure solver needs to lookup a type of their
originator and based on the wrapper type compute and
assign types to projection and/or wrapper.

Resolves: https://github.com/apple/swift/issues/59295
Resolves: rdar://94506352

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/CSClosure.cpp [hunk 1]
```diff
@@ -96,6 +96,29 @@ class TypeVariableRefFinder : public ASTWalker {
       if (!var)
         return {true, expr};
 
+      if (auto *wrappedVar = var->getOriginalWrappedProperty()) {
+        auto outermostWrapperAttr =
+            wrappedVar->getAttachedPropertyWrappers().back();
```
Comment (hborla) [inline lib/Sema/CSClosure.cpp:101]: Hmm, `getAttachedPropertyWrappers().front()` should give the outermost wrapper attribute

File: validation-test/Sema/type_checker_crashers_fixed/issue59294.swift [hunk 2]
```diff
@@ -25,18 +26,94 @@ struct WrapperValue<Value> {
...
+  static func testNestedWrappers() {
+    func test(_: (Bool) -> Void) {}
+    func test(_: () -> Void) {}
+
+    test {
+      if true {
+        @OuterWrapper @WrapperValue var value = $0
+        if true {
+          let _: Bool = _value == $0
+          let _: OuterWrapper<Bool> = $value
```
Comment (hborla) [inline validation-test/Sema/type_checker_crashers_fixed/issue59294.swift:103]: The type of the projected value here should be `OuterWrapper<WrapperValue<Bool>>`

File: lib/Sema/CSClosure.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSClosure.cpp:101]: Indeed, I thought that it was in reverse, going to add a method to make it harder to get wrong...

Comment (xedin) [reply lib/Sema/CSClosure.cpp]: Fixed!

File: validation-test/Sema/type_checker_crashers_fixed/issue59294.swift (hunk 2 above)
Comment (xedin) [reply validation-test/Sema/type_checker_crashers_fixed/issue59294.swift]: Fixed by using correct "outer" wrapper, thank you!

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**